// ng-form-rules/esm2015/lib/form-rules/models/abstract-model-settings.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { ModelSettingsBuilder } from "../helper/model-settings-builder";
/**
 * Base class for model settings
 * @abstract
 * @template T
 */
export class AbstractModelSettings {
    /**
     * Creates model settings
     * @param {?} name Name of the model settings
     */
    constructor(name) {
        this.builder = new ModelSettingsBuilder();
        this._name = name;
        this.properties = this.buildProperties() || [];
    }
    /**
     * Name of the model setting
     * @return {?}
     */
    get name() {
        return this._name;
    }
}
function AbstractModelSettings_tsickle_Closure_declarations() {
    /** @type {?} */
    AbstractModelSettings.prototype._name;
    /**
     * Helper for building properties, tests, and rules
     * @type {?}
     */
    AbstractModelSettings.prototype.builder;
    /**
     * Properties configured for the model settings
     * @type {?}
     */
    AbstractModelSettings.prototype.properties;
    /**
     * Abstract function for building the properties
     * @abstract
     * @return {?}
     */
    AbstractModelSettings.prototype.buildProperties = function () { };
}
