// ng-form-rules/esm5/lib/form-rules/models/property-base.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * Base class for a property
 * @abstract
 * @template T
 */
var /**
 * Base class for a property
 * @abstract
 * @template T
 */
PropertyBase = /** @class */ (function () {
    function PropertyBase() {
        this._dependencyPropertySubscriptions = [];
        /**
         * Validation tests
         */
        this.valid = [];
        /**
         * Editability tests
         */
        this.edit = [];
        /**
         * The event name for control to update upon.
         */
        this.updateOn = 'change';
        /**
         * Options for how to respond to value changes on property control and dependency controls
         */
        this.valueChangeOptions = {
            dependencyProperties: {
                valid: { distinctUntilChanged: false, debounceMilliseconds: 0 },
                edit: { distinctUntilChanged: false, debounceMilliseconds: 0 }
            },
            self: {
                asyncValid: { distinctUntilChanged: false, debounceMilliseconds: 0 },
                edit: { distinctUntilChanged: false, debounceMilliseconds: 0 }
            }
        };
    }
    Object.defineProperty(PropertyBase.prototype, "absolutePath", {
        /**
         * Absolute path to the property from root
         */
        get: /**
         * Absolute path to the property from root
         * @return {?}
         */
        function () {
            return this._absolutePath;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PropertyBase.prototype, "dependencyPropertySubscriptions", {
        /**
         * Absolute path to property from root
         */
        get: /**
         * Absolute path to property from root
         * @return {?}
         */
        function () {
            return this._dependencyPropertySubscriptions;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PropertyBase.prototype, "ownerModelSettingsName", {
        /**
         * Absolute path to property from root
         */
        get: /**
         * Absolute path to property from root
         * @return {?}
         */
        function () {
            return this._ownerModelSettingsName;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Determines if the property is an ArrayItemProperty
     * @param property Property to check
     * @returns Whether or not the property is an ArrayItemProperty
     */
    /**
     * Determines if the property is an ArrayItemProperty
     * @template T
     * @param {?} property Property to check
     * @return {?} Whether or not the property is an ArrayItemProperty
     */
    PropertyBase.isArrayItemProperty = /**
     * Determines if the property is an ArrayItemProperty
     * @template T
     * @param {?} property Property to check
     * @return {?} Whether or not the property is an ArrayItemProperty
     */
    function (property) {
        return !(/** @type {?} */ (property)).name;
    };
    /**
     * DO NOT CALL! Internal use only
     * @param absolutePath Absolute path to the property from the root
     */
    /**
     * DO NOT CALL! Internal use only
     * @param {?} absolutePath Absolute path to the property from the root
     * @return {?}
     */
    PropertyBase.prototype.setAbsolutePath = /**
     * DO NOT CALL! Internal use only
     * @param {?} absolutePath Absolute path to the property from the root
     * @return {?}
     */
    function (absolutePath) {
        this._absolutePath = absolutePath;
    };
    /**
     * DO NOT CALL! Internal use only
     * @param subscription Subscription for the dependency property
     */
    /**
     * DO NOT CALL! Internal use only
     * @param {?} subscription Subscription for the dependency property
     * @return {?}
     */
    PropertyBase.prototype.addDependencyPropertySubscription = /**
     * DO NOT CALL! Internal use only
     * @param {?} subscription Subscription for the dependency property
     * @return {?}
     */
    function (subscription) {
        this._dependencyPropertySubscriptions.push(subscription);
    };
    /**
     * DO NOT CALL! Internal use only
     */
    /**
     * DO NOT CALL! Internal use only
     * @return {?}
     */
    PropertyBase.prototype.clearDependencyPropertySubscriptions = /**
     * DO NOT CALL! Internal use only
     * @return {?}
     */
    function () {
        this.dependencyPropertySubscriptions.forEach(function (sub$) { return sub$.unsubscribe(); });
        this._dependencyPropertySubscriptions = [];
    };
    /**
     * DO NOT CALL! Internal use only
     * @param name Name of the owner model settings
     */
    /**
     * DO NOT CALL! Internal use only
     * @param {?} name Name of the owner model settings
     * @return {?}
     */
    PropertyBase.prototype.setOwnerModelSettingsName = /**
     * DO NOT CALL! Internal use only
     * @param {?} name Name of the owner model settings
     * @return {?}
     */
    function (name) {
        this._ownerModelSettingsName = name;
    };
    return PropertyBase;
}());
/**
 * Base class for a property
 * @abstract
 * @template T
 */
export { PropertyBase };
function PropertyBase_tsickle_Closure_declarations() {
    /** @type {?} */
    PropertyBase.prototype._absolutePath;
    /** @type {?} */
    PropertyBase.prototype._dependencyPropertySubscriptions;
    /** @type {?} */
    PropertyBase.prototype._ownerModelSettingsName;
    /**
     * Validation tests
     * @type {?}
     */
    PropertyBase.prototype.valid;
    /**
     * Editability tests
     * @type {?}
     */
    PropertyBase.prototype.edit;
    /**
     * Properties for a complex object
     * @type {?}
     */
    PropertyBase.prototype.properties;
    /**
     * Property for an item of an array
     * @type {?}
     */
    PropertyBase.prototype.arrayItemProperty;
    /**
     * The event name for control to update upon.
     * @type {?}
     */
    PropertyBase.prototype.updateOn;
    /**
     * Options for how to respond to value changes on property control and dependency controls
     * @type {?}
     */
    PropertyBase.prototype.valueChangeOptions;
}
