// ng-form-rules/esm5/lib/form-rules/helper/model-settings-builder.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Property } from '../models/property';
import { ArrayItemProperty } from '../models/array-item-property';
import { ResultsPassRequirement } from '../models/results-pass-requirement';
/**
 * Utility class for building model settings
 */
var /**
 * Utility class for building model settings
 */
ModelSettingsBuilder = /** @class */ (function () {
    function ModelSettingsBuilder() {
    }
    /**
     * Creates a property
     * @param propertyName Property name
     * @param extend Function to extend the property
     * @returns Created property
     */
    /**
     * Creates a property
     * @template T
     * @param {?} propertyName Property name
     * @param {?=} extend Function to extend the property
     * @return {?} Created property
     */
    ModelSettingsBuilder.prototype.property = /**
     * Creates a property
     * @template T
     * @param {?} propertyName Property name
     * @param {?=} extend Function to extend the property
     * @return {?} Created property
     */
    function (propertyName, extend) {
        if (!propertyName || typeof propertyName !== 'string')
            throw Error('Invalid property name');
        var /** @type {?} */ property = new Property(propertyName);
        if (extend)
            extend(property);
        return property;
    };
    /**
     * Creates an array item property
     * @param extend Function to extend the array item property
     * @returns Created array item property
     */
    /**
     * Creates an array item property
     * @template T
     * @param {?=} extend Function to extend the array item property
     * @return {?} Created array item property
     */
    ModelSettingsBuilder.prototype.arrayItemProperty = /**
     * Creates an array item property
     * @template T
     * @param {?=} extend Function to extend the array item property
     * @return {?} Created array item property
     */
    function (extend) {
        var /** @type {?} */ property = new ArrayItemProperty();
        if (extend)
            extend(property);
        return property;
    };
    /**
     * Creates a validation test
     * @param message Message when the test fails
     * @param check Check rule set for the test
     * @param condition Condition rule set for the test
     * @returns Created validation test
     */
    /**
     * Creates a validation test
     * @template T
     * @param {?} message Message when the test fails
     * @param {?} check Check rule set for the test
     * @param {?=} condition Condition rule set for the test
     * @return {?} Created validation test
     */
    ModelSettingsBuilder.prototype.validTest = /**
     * Creates a validation test
     * @template T
     * @param {?} message Message when the test fails
     * @param {?} check Check rule set for the test
     * @param {?=} condition Condition rule set for the test
     * @return {?} Created validation test
     */
    function (message, check, condition) {
        return this.validNamedTest(null, message, check, condition);
    };
    /**
     * Creates a named validation test
     * @param name Name of the test
     * @param message Message when the test fails
     * @param check Check rule set for the test
     * @param condition Condition rule set for the test
     * @returns Created named validation test
     */
    /**
     * Creates a named validation test
     * @template T
     * @param {?} name Name of the test
     * @param {?} message Message when the test fails
     * @param {?} check Check rule set for the test
     * @param {?=} condition Condition rule set for the test
     * @return {?} Created named validation test
     */
    ModelSettingsBuilder.prototype.validNamedTest = /**
     * Creates a named validation test
     * @template T
     * @param {?} name Name of the test
     * @param {?} message Message when the test fails
     * @param {?} check Check rule set for the test
     * @param {?=} condition Condition rule set for the test
     * @return {?} Created named validation test
     */
    function (name, message, check, condition) {
        return /** @type {?} */ ({
            name: name,
            message: message,
            check: check,
            condition: condition
        });
    };
    /**
     * Creates an edit test
     * @param check Check rule set for the test
     * @param condition Condition rule set for the test
     * @returns Created edit test
     */
    /**
     * Creates an edit test
     * @template T
     * @param {?} check Check rule set for the test
     * @param {?=} condition Condition rule set for the test
     * @return {?} Created edit test
     */
    ModelSettingsBuilder.prototype.editTest = /**
     * Creates an edit test
     * @template T
     * @param {?} check Check rule set for the test
     * @param {?=} condition Condition rule set for the test
     * @return {?} Created edit test
     */
    function (check, condition) {
        return this.editNamedTest(null, check, condition);
    };
    /**
     * Creates a named edit test
     * @param name Name of the test
     * @param check Check rule set for the test
     * @param condition Condition rule set for the test
     * @returns Created named edit test
     */
    /**
     * Creates a named edit test
     * @template T
     * @param {?} name Name of the test
     * @param {?} check Check rule set for the test
     * @param {?=} condition Condition rule set for the test
     * @return {?} Created named edit test
     */
    ModelSettingsBuilder.prototype.editNamedTest = /**
     * Creates a named edit test
     * @template T
     * @param {?} name Name of the test
     * @param {?} check Check rule set for the test
     * @param {?=} condition Condition rule set for the test
     * @return {?} Created named edit test
     */
    function (name, check, condition) {
        return /** @type {?} */ ({
            name: name,
            check: check,
            condition: condition
        });
    };
    /**
     * Creates a rule
     * @param func Function for the rule
     * @param options Additional rule options
     * @returns Created rule
     */
    /**
     * Creates a rule
     * @template T, R
     * @param {?} func Function for the rule
     * @param {?=} options Additional rule options
     * @return {?} Created rule
     */
    ModelSettingsBuilder.prototype.rule = /**
     * Creates a rule
     * @template T, R
     * @param {?} func Function for the rule
     * @param {?=} options Additional rule options
     * @return {?} Created rule
     */
    function (func, options) {
        return this.ruleCombo(func, null, options);
    };
    /**
     * Creates an async rule
     * @param asyncFunc Async function for the rule
     * @param options Additional rule options
     * @returns Created rule
     */
    /**
     * Creates an async rule
     * @template T, R
     * @param {?} asyncFunc Async function for the rule
     * @param {?=} options Additional rule options
     * @return {?} Created rule
     */
    ModelSettingsBuilder.prototype.ruleAsync = /**
     * Creates an async rule
     * @template T, R
     * @param {?} asyncFunc Async function for the rule
     * @param {?=} options Additional rule options
     * @return {?} Created rule
     */
    function (asyncFunc, options) {
        return this.ruleCombo(null, asyncFunc, options);
    };
    /**
     * Creates a combo rule that has both syncronous and async functions
     * @param func Function for the rule
     * @param asyncFunc Async function for the rule
     * @param options Additional rule options
     * @returns Created rule
     */
    /**
     * Creates a combo rule that has both syncronous and async functions
     * @template T, R
     * @param {?} func Function for the rule
     * @param {?} asyncFunc Async function for the rule
     * @param {?=} options Additional rule options
     * @return {?} Created rule
     */
    ModelSettingsBuilder.prototype.ruleCombo = /**
     * Creates a combo rule that has both syncronous and async functions
     * @template T, R
     * @param {?} func Function for the rule
     * @param {?} asyncFunc Async function for the rule
     * @param {?=} options Additional rule options
     * @return {?} Created rule
     */
    function (func, asyncFunc, options) {
        return /** @type {?} */ ({
            func: func,
            asyncFunc: asyncFunc,
            options: options
        });
    };
    /**
     * Creates a rule group
     * @param ruleSets Rule sets for the rule group
     * @param options Additional rule group options
     * @returns Created rule group
     */
    /**
     * Creates a rule group
     * @template T
     * @param {?} ruleSets Rule sets for the rule group
     * @param {?=} options Additional rule group options
     * @return {?} Created rule group
     */
    ModelSettingsBuilder.prototype.ruleGroup = /**
     * Creates a rule group
     * @template T
     * @param {?} ruleSets Rule sets for the rule group
     * @param {?=} options Additional rule group options
     * @return {?} Created rule group
     */
    function (ruleSets, options) {
        return /** @type {?} */ ({
            rules: ruleSets,
            options: options || { resultRequirements: ResultsPassRequirement.AllPass }
        });
    };
    return ModelSettingsBuilder;
}());
/**
 * Utility class for building model settings
 */
export { ModelSettingsBuilder };
